crates/matrix-sdk/src/client/builder.rs

```diff
@@ -363,71 +395,82 @@ impl ClientBuilder {
-
-        let homeserver = match homeserver_cfg {
-            HomeserverConfig::Url(url) => {
-                #[cfg(feature = "experimental-sliding-sync")]
-                {
-                    sliding_sync_proxy =
-                        self.sliding_sync_proxy.as_ref().map(|url| Url::parse(url)).transpose()?;
+            HomeserverConfig::ServerName { server: server_name, protocol } => {
+                let well_known = discover_homeserver(server_name, protocol, &http_client).await?;
+                (well_known.homeserver.base_url.clone(), Some(well_known))
+            }
+
+            HomeserverConfig::ServerNameOrUrl(server_name_or_url) => {
+                // Store the result to return at the end. If this doesn't get modified, then the
+                // supplied name is neither a server name, nor a valid URL.
+                let mut homeserver_details: Option<(
+                    String,
+                    Option<discover_homeserver::Response>,
+                )> = None;
+                let mut discovery_error: Option<ClientBuildError> = None;
+
+                // Attempt discovery as a server name first.
+                let sanitize_result = sanitize_server_name(&server_name_or_url);
+                if let Ok(server_name) = sanitize_result.as_ref() {
+                    let protocol = if server_name_or_url.starts_with("http://") {
+                        UrlScheme::Http
+                    } else {
+                        UrlScheme::Https
+                    };
+
+                    match discover_homeserver(server_name.clone(), protocol, &http_client).await {
+                        Ok(well_known) => {
+                            homeserver_details =
+                                Some((well_known.homeserver.base_url.clone(), Some(well_known)));
+                        }
+                        Err(e) => {
+                            debug!(error = ?e, "Well-known discovery failed.");
+                            discovery_error = Some(e);
+                        }
+                    }
                 }
 
-                #[cfg(feature = "experimental-oidc")]
-                {
-                    allow_insecure_oidc = url.starts_with("http://");
+                // When discovery fails, or the input isn't a valid server name, fallback to
+                // trying a homeserver URL if supplied.
+                if homeserver_details.is_none() {
+                    if let Ok(homeserver_url) = Url::parse(&server_name_or_url) {
+                        // Make sure the URL is definitely for a homeserver.
+                        if check_is_homeserver(&homeserver_url, &http_client).await {
+                            homeserver_details = Some((homeserver_url.to_string(), None));
+                        }
+                    }
                 }
 
-                url
+                homeserver_details
+                    .ok_or(discovery_error.unwrap_or(ClientBuildError::InvalidServerName))?
```

Comment:
It's a bit unfortunate that we have to choose an error, when it's really both errors happening at the same time.

Instead, could we have the `InvalidServerName` error have an extra data field: `discovery_error` (and it would always be set, not Optional).

Reply:
I would disagree with this:
- If `sanitize_server_name` and `Url::parse(&server_name_or_url)` both fail we never try discovery, the server name is invalid.
- If the first is successful then there will be a discovery error and the server name is valid.
- And if the server name sanitisation fails, but it is a valid URL, something has gone wrong and we have a bug in `sanitize_server_name`.